lib/ruby_smb/server/share/provider/virtual_disk.rb

```diff
@@ -0,0 +1,115 @@
+          #
+          # @param [String] path The path of the file to add, relative to the share root.
+          # @param [String, Pathname] mapped_path The path on the local file system to map into the virtual file system.
+          def add_mapped_file(path, mapped_path)
+            path = VirtualPathname.cleanpath(path)
+            path = File::SEPARATOR + path unless path.start_with?(File::SEPARATOR)
+
+            vf = VirtualMappedFile.new(self, path, mapped_path)
+            add(vf)
+          end
+
+          # Add a static file to the virtual file system. A static file is one whose contents are known at creation time
+          # and do not change.
+          #
+          # @param [String] path The path of the file to add, relative to the share root.
+          # @param [String, #read] content The static content to add.
+          # @param [File::Stat] stat An explicit stat object describing the file.
+          def add_static_file(path, content, stat: nil)
+            path = VirtualPathname.cleanpath(path)
+            path = File::SEPARATOR + path unless path.start_with?(File::SEPARATOR)
+            raise ArgumentError.new('must be a file') if stat && !stat.file?
+
+            content = content.read if content.respond_to?(:read)
+            vf = VirtualStaticFile.new(self, path, content, stat: stat)
+            add(vf)
+          end
+
+          # Add a file object to the virtual file system. This will read the file and create a new static file from the
+          # contents. The file's #stat object will be copied over.
+          #
+          # @param [String] path The path of the file to add, relative to the share root.
+          # @param [#stat] file_obj The file object to take the stat object and content from.
+          def add_static_fileobj(path, file_obj)
+            add_static_file(path, file_obj, stat: file_obj.stat)
+          end
+
+          def add(virtual_pathname)
+            raise ArgumentError.new('paths must be absolute') unless virtual_pathname.absolute?
+
+            path = virtual_pathname.to_s
+            raise ArgumentError.new('paths must be normalized') unless VirtualPathname.cleanpath(path) == path
+
+            path_parts = path.split(VirtualPathname::SEPARATOR)
+            2.upto(path_parts.length - 1) do |idx|
+              ancestor = path_parts[0...idx].join(path[VirtualPathname::SEPARATOR])
+              next if @vfs[ancestor]&.directory?
+
+              @vfs[ancestor] = VirtualPathname.new(self, ancestor, stat: VirtualStat.new(directory?: true))
+            end
+
+            @vfs[path] = virtual_pathname
+          end
+
+          def new_processor(server_client, session)
+            scoped_virtual_disk = self.class.new(@name)
+            @vfs.each_value do |path|
+              path = path.dup
+              path.virtual_disk = scoped_virtual_disk if path.is_a?(VirtualPathname)
+              path = path.generate(server_client, session) if path.is_a?(VirtualDynamicFile)
```

Comment:
So I originally had it that way. IIRC there's still an implementation in the commit history that instead calculated it when the file was read. When I started tying this into Metasploit, its main limitation become apparent and that was the file size. If you know and keep the file size static, then everything will be fine. Alot of clients will query the file information before opening it and reading it and I think it's best if we keep the size consistent so the client can query the size, open the file, and read all of its contents without the size changing between those steps. Generating it on tree-connect guarantees that the file's information can't be accessed until it's contents are generated and locked into place. This removes the need to know the size ahead of time which was also problematic because you'd then need to pad or truncate the dynamically generated contents.

Happy to resurrect that pattern, do you have a use case for it in mind?

Reply:
It makes totally sense. I'm wondering when the clients usually check the file size when they want to read or write to it.

In my understanding, and it is pure speculation since I've not tested it, it would make sense to check the size as soon as the file is open (SMB1 `NtCreateAndx` or SMB2/3 `Create`), read it as many time as needed with a fixed size and close it. The file might change between two open/read/close cycles and checking the size when opening it again makes sense to me. It looks like Windows client behaves like that. It keeps the Tree open and only close it when explicitly asked (`net use \\1.2.3.4\public /delete`). I'm not sure how other clients behave though.

One use case would be a C2 server that dispatches commands using SMB. The payload on the target could get the commands from the server using a simple read file, which would be implemented as a virtual dynamic file.